File: LC 1696.cpp
class Solution {
public:
    int maxResult(vector<int>& nums, int k) {
        int n = nums.size();
        int ans = 0;
        vector<int> dp(n, INT_MIN);
        dp[0] = nums[0];
        deque<int> dq{0};
        for (int i = 1; i < n; ++i) {
            while (i - dq.front() > k) {
                dq.pop_front();
            }
            dp[i] = dp[dq.front()] + nums[i];
            while (!dq.empty() && dp[dq.back()] <= dp[i]) {
                dq.pop_back();
            }
            dq.push_back(i);
        }
        return dp[n-1];
    }
};

File: 2000/LC 1696.go
func maxResult(nums []int, k int) int {
    n := len(nums)
    f := make([]int, n)
    f[0] = nums[0]
    q := []int{0}
    for i := 1; i < n; i++ {
        for len(q) > 0 && i - q[0] > k {
            q = q[1:]
        }
        f[i] = f[q[0]] + nums[i]
        for len(q) > 0 && f[q[len(q)-1]] < f[i] {
            q = q[:len(q)-1]
        }
        q = append(q, i)
    }

    return f[n-1]
}